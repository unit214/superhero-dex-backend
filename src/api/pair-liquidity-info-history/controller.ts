import {
  Controller,
  Get,
  ParseEnumPipe,
  ParseIntPipe,
  Query,
} from '@nestjs/common';
import { PairLiquidityInfoHistoryService } from './service';
import { ApiOperation, ApiQuery, ApiResponse } from '@nestjs/swagger';
import * as dto from '../../dto';
import { OrderQueryEnum } from '../../dto';
import { ContractAddress } from '../../lib/utils';

@Controller('history/liquidity')
export class PairLiquidityInfoHistoryController {
  constructor(
    private readonly pairLiquidityInfoHistoryService: PairLiquidityInfoHistoryService,
  ) {}

  @Get()
  @ApiOperation({
    summary: 'Retrieve all entries of the pair liquidity info history',
    deprecated: true,
  })
  @ApiQuery({
    name: 'limit',
    type: Number,
    description: 'Limit of history entries per page (default: 100, max: 100)',
    required: false,
  })
  @ApiQuery({
    name: 'offset',
    type: Number,
    description: 'Offset of page (default: 0)',
    required: false,
  })
  @ApiQuery({
    name: 'order',
    enum: OrderQueryEnum,
    description:
      'Sorts history entries in ascending or descending order (default: asc)',
    required: false,
  })
  @ApiQuery({
    name: 'pair-address',
    type: String,
    description: 'Retrieve only history entries for the given pair address',
    required: false,
  })
  @ApiQuery({
    name: 'height',
    type: Number,
    description: 'Retrieve only history entries for the given height',
    required: false,
  })
  @ApiQuery({
    name: 'from-block-time',
    type: Number,
    description:
      'Retrieve only history entries that are equal or newer than the given micro block time',
    required: false,
  })
  @ApiQuery({
    name: 'to-block-time',
    type: Number,
    description:
      'Retrieve only history entries that are equal or older than the given micro block time',
    required: false,
  })
  @ApiResponse({ status: 200, type: [dto.PairLiquidityInfoHistoryEntry] })
  async findAll(
    @Query('limit', new ParseIntPipe({ optional: true })) limit: number = 100,
    @Query('offset', new ParseIntPipe({ optional: true })) offset: number = 0,
    @Query('order', new ParseEnumPipe(OrderQueryEnum, { optional: true }))
    order: OrderQueryEnum = OrderQueryEnum.asc,
    @Query('pair-address') pairAddress?: ContractAddress,
    @Query('height', new ParseIntPipe({ optional: true })) height?: number,
    @Query('from-block-time', new ParseIntPipe({ optional: true }))
    fromBlockTime?: number,
    @Query('to-block-time', new ParseIntPipe({ optional: true }))
    toBlockTime?: number,
  ): Promise<dto.PairLiquidityInfoHistoryEntry[]> {
    return this.pairLiquidityInfoHistoryService
      .getAllHistoryEntries(
        Number(limit),
        Number(offset),
        order,
        pairAddress,
        height ? Number(height) : undefined,
        fromBlockTime ? BigInt(fromBlockTime) : undefined,
        toBlockTime ? BigInt(toBlockTime) : undefined,
      )
      .then((entries) =>
        entries.map((entry) => ({
          pairAddress: entry.pair.address,
          liquidityInfo: {
            totalSupply: entry.totalSupply,
            reserve0: entry.reserve0,
            reserve1: entry.reserve1,
          },
          height: entry.height,
          microBlockHash: entry.microBlockHash,
          microBlockTime: entry.microBlockTime.toString(),
        })),
      );
  }
}
