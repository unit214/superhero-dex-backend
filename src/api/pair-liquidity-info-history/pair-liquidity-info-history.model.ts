import { Pair, PairLiquidityInfoHistory, Token } from '@prisma/client';

export type PairLiquidityInfoHistoryWithTokens = {
  pair: Pair & { token0: Token; token1: Token };
} & PairLiquidityInfoHistory;
