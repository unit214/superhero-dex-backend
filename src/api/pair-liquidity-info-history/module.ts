import { Module } from '@nestjs/common';
import { PairLiquidityInfoHistoryService } from './service';
import { PairLiquidityInfoHistoryController } from './controller';
import { DatabaseModule } from '../../database/database.module';

@Module({
  imports: [DatabaseModule],
  controllers: [PairLiquidityInfoHistoryController],
  providers: [PairLiquidityInfoHistoryService],
})
export class PairLiquidityInfoHistoryModule {}
