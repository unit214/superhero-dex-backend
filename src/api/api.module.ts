import { Module } from '@nestjs/common';

import { GraphController } from '@/api/graph/graph.controller';
import { GraphService } from '@/api/graph/graph.service';
import { PairLiquidityInfoHistoryController } from '@/api/pair-liquidity-info-history/pair-liquidity-info-history.controller';
import { PairLiquidityInfoHistoryService } from '@/api/pair-liquidity-info-history/pair-liquidity-info-history.service';
import { PairsController } from '@/api/pairs/pairs.controller';
import { PairsService } from '@/api/pairs/pairs.service';
import { SwapRoutesService } from '@/api/swap-routes/swap-route.service';
import { SwapRoutesController } from '@/api/swap-routes/swap-routes.controller';
import { TokensController } from '@/api/tokens/tokens.controller';
import { TokensService } from '@/api/tokens/tokens.service';
import { DatabaseModule } from '@/database/database.module';

@Module({
  imports: [DatabaseModule],
  controllers: [
    PairLiquidityInfoHistoryController,
    PairsController,
    TokensController,
    SwapRoutesController,
    GraphController,
  ],
  providers: [
    PairLiquidityInfoHistoryService,
    PairsService,
    TokensService,
    SwapRoutesService,
    GraphService,
  ],
})
export class ApiModule {}
