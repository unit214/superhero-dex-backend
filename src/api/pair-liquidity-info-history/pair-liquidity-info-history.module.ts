import { Module } from '@nestjs/common';
import { PairLiquidityInfoHistoryService } from './pair-liquidity-info-history.service';
import { PairLiquidityInfoHistoryController } from './pair-liquidity-info-history.controller';
import { DatabaseModule } from '../../database/database.module';

@Module({
  imports: [DatabaseModule],
  controllers: [PairLiquidityInfoHistoryController],
  providers: [PairLiquidityInfoHistoryService],
})
export class PairLiquidityInfoHistoryModule {}
