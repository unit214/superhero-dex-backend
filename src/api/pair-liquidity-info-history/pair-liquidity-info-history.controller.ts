import {
  Controller,
  Get,
  ParseEnumPipe,
  ParseIntPipe,
  Query,
} from '@nestjs/common';
import { ApiOperation, ApiQuery, ApiResponse } from '@nestjs/swagger';

import { OrderQueryEnum, PairLiquidityInfoHistoryEntry } from '@/api/api.model';
import { PairLiquidityInfoHistoryService } from '@/api/pair-liquidity-info-history/pair-liquidity-info-history.service';
import { ContractAddress } from '@/clients/sdk-client.model';

@Controller('history/liquidity')
export class PairLiquidityInfoHistoryController {
  constructor(
    private readonly pairLiquidityInfoHistoryService: PairLiquidityInfoHistoryService,
  ) {}

  @Get()
  @ApiOperation({
    summary: 'Retrieve all entries of the pair liquidity info history',
  })
  @ApiQuery({
    name: 'limit',
    type: Number,
    description: 'Limit of history entries per page (default: 100, max: 100)',
    required: false,
  })
  @ApiQuery({
    name: 'offset',
    type: Number,
    description: 'Offset of page (default: 0)',
    required: false,
  })
  @ApiQuery({
    name: 'order',
    enum: OrderQueryEnum,
    description:
      'Sorts history entries in ascending or descending order (default: asc)',
    required: false,
  })
  @ApiQuery({
    name: 'pairAddress',
    type: String,
    description: 'Retrieve only history entries for the given pair address',
    required: false,
  })
  @ApiQuery({
    name: 'height',
    type: Number,
    description: 'Retrieve only history entries for the given height',
    required: false,
  })
  @ApiQuery({
    name: 'fromBlockTime',
    type: Number,
    description:
      'Retrieve only history entries that are equal or newer than the given micro block time',
    required: false,
  })
  @ApiQuery({
    name: 'toBlockTime',
    type: Number,
    description:
      'Retrieve only history entries that are equal or older than the given micro block time',
    required: false,
  })
  @ApiResponse({ status: 200, type: [PairLiquidityInfoHistoryEntry] })
  async findAll(
    @Query('limit', new ParseIntPipe({ optional: true })) limit: number = 100,
    @Query('offset', new ParseIntPipe({ optional: true })) offset: number = 0,
    @Query('order', new ParseEnumPipe(OrderQueryEnum, { optional: true }))
    order: OrderQueryEnum = OrderQueryEnum.asc,
    @Query('pairAddress') pairAddress?: ContractAddress,
    @Query('height', new ParseIntPipe({ optional: true })) height?: number,
    @Query('fromBlockTime', new ParseIntPipe({ optional: true }))
    fromBlockTime?: number,
    @Query('toBlockTime', new ParseIntPipe({ optional: true }))
    toBlockTime?: number,
  ): Promise<PairLiquidityInfoHistoryEntry[]> {
    return this.pairLiquidityInfoHistoryService
      .getAllHistoryEntries(
        Number(limit),
        Number(offset),
        order,
        pairAddress,
        height != null ? Number(height) : undefined,
        fromBlockTime != null ? BigInt(fromBlockTime) : undefined,
        toBlockTime != null ? BigInt(toBlockTime) : undefined,
      )
      .then((entries) =>
        entries.map((entry) => ({
          pairAddress: entry.pair.address,
          type: entry.eventType,
          reserve0: entry.reserve0.toString(),
          reserve1: entry.reserve1.toString(),
          deltaReserve0: entry.deltaReserve0.toString(),
          deltaReserve1: entry.deltaReserve1.toString(),
          aeUsdPrice: entry.aeUsdPrice.toString(),
          height: entry.height,
          microBlockHash: entry.microBlockHash,
          microBlockTime: entry.microBlockTime.toString(),
          transactionHash: entry.transactionHash,
          transactionIndex: entry.transactionIndex.toString(),
          logIndex: entry.logIndex,
        })),
      );
  }
}
