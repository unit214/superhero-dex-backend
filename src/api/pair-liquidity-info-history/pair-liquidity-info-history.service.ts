import { Injectable } from '@nestjs/common';
import BigNumber from 'bignumber.js';

import { PairLiquidityInfoHistoryEntry } from '@/api/api.model';
import { PairLiquidityInfoHistoryWithTokens } from '@/api/pair-liquidity-info-history/pair-liquidity-info-history.model';
import { PairLiquidityInfoHistoryDbService } from '@/database/pair-liquidity-info-history/pair-liquidity-info-history-db.service';
import { calculateUsdValue } from '@/lib/utils';

@Injectable()
export class PairLiquidityInfoHistoryService {
  constructor(
    private readonly pairLiquidityInfoHistoryDb: PairLiquidityInfoHistoryDbService,
  ) {}

  getAllHistoryEntries = this.pairLiquidityInfoHistoryDb.getAll;

  mapToEntryWithPrice(
    entry: PairLiquidityInfoHistoryWithTokens,
  ): PairLiquidityInfoHistoryEntry {
    const usdItems: {
      reserve0Usd: string | null;
      reserve1Usd: string | null;
      delta0UsdValue: string | null;
      delta1UsdValue: string | null;
      txUsdFee: string | null;
    } = {
      reserve0Usd: null,
      reserve1Usd: null,
      delta0UsdValue: null,
      delta1UsdValue: null,
      txUsdFee: null,
    };
    const token0AePrice =
      entry.token0AePrice === null || entry.token0AePrice.toString() === '-1'
        ? null
        : entry.token0AePrice?.toString();

    const token1AePrice =
      entry.token1AePrice === null || entry.token1AePrice.toString() === '-1'
        ? null
        : entry.token1AePrice?.toString();

    if (token0AePrice !== null && token1AePrice !== null) {
      usdItems.reserve0Usd = calculateUsdValue({
        reserve: entry.reserve0.toString(),
        tokenAePrice: token0AePrice,
        decimals: entry.pair.token0.decimals,
        aeUsdPrice: entry.aeUsdPrice.toString(),
      });
      usdItems.reserve1Usd = calculateUsdValue({
        reserve: entry.reserve1.toString(),
        tokenAePrice: token1AePrice,
        decimals: entry.pair.token1.decimals,
        aeUsdPrice: entry.aeUsdPrice.toString(),
      });
      usdItems.delta0UsdValue = calculateUsdValue({
        reserve: entry.deltaReserve0.toString(),
        tokenAePrice: token0AePrice,
        decimals: entry.pair.token0.decimals,
        aeUsdPrice: entry.aeUsdPrice.toString(),
      });
      usdItems.delta1UsdValue = calculateUsdValue({
        reserve: entry.deltaReserve1.toString(),
        tokenAePrice: token1AePrice,
        decimals: entry.pair.token1.decimals,
        aeUsdPrice: entry.aeUsdPrice.toString(),
      });
      usdItems.txUsdFee = new BigNumber(usdItems.delta0UsdValue)
        .multipliedBy(0.003)
        .toString();
    }

    return {
      pairAddress: entry.pair.address,
      senderAccount: entry.senderAccount,
      type: entry.eventType,
      reserve0: entry.reserve0.toString(),
      reserve1: entry.reserve1.toString(),
      deltaReserve0: entry.deltaReserve0.toString(),
      deltaReserve1: entry.deltaReserve1.toString(),
      token0AePrice: token0AePrice,
      token1AePrice: token1AePrice,
      aeUsdPrice: entry.aeUsdPrice.toString(),
      height: entry.height,
      microBlockHash: entry.microBlockHash,
      microBlockTime: entry.microBlockTime.toString(),
      transactionHash: entry.transactionHash,
      transactionIndex: entry.transactionIndex.toString(),
      logIndex: entry.logIndex,
      ...usdItems,
    };
  }
}
