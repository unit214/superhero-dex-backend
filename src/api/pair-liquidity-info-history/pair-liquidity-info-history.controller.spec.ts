import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { Pair, PairLiquidityInfoHistory } from '@prisma/client';
import * as request from 'supertest';

import { OrderQueryEnum } from '@/api/api.model';
import { PairLiquidityInfoHistoryController } from '@/api/pair-liquidity-info-history/pair-liquidity-info-history.controller';
import { PairLiquidityInfoHistoryService } from '@/api/pair-liquidity-info-history/pair-liquidity-info-history.service';

const mockPairLiquidityInfoHistoryService = {
  getAllHistoryEntries: jest.fn(),
};

describe('PairLiquidityInfoHistoryController', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [PairLiquidityInfoHistoryController],
      providers: [
        {
          provide: PairLiquidityInfoHistoryService,
          useValue: mockPairLiquidityInfoHistoryService,
        },
      ],
    }).compile();

    app = module.createNestApplication();
    await app.init();
  });

  describe('GET /history/liquidity', () => {
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should return history entries and use default values for empty params', async () => {
      // Mocks
      const historyEntry1: { pair: Pair } & PairLiquidityInfoHistory = {
        id: 1,
        pairId: 1,
        totalSupply: '2000148656239820912122563',
        reserve0: '950875688379385634428666',
        reserve1: '4208476309359648851631167',
        height: 912485,
        microBlockHash: 'mh_Tx43Gh3acudUNSUWihPcV1Se4XcoFK3aUFAtFZk2Z4Zv7igZs',
        microBlockTime: 1709027642807n,
        updatedAt: new Date('2024-03-20 17:04:51.625'),
        pair: {
          id: 1,
          address: 'ct_efYtiwDg4YZxDWE3iLPzvrjb92CJPvzGwriv4ZRuvuTDMNMb9',
          t0: 15,
          t1: 5,
          synchronized: true,
        },
      };

      const historyEntry2: { pair: Pair } & PairLiquidityInfoHistory = {
        id: 2,
        pairId: 3,
        totalSupply: '9954575303087659158151',
        reserve0: '20210309618736130321327',
        reserve1: '4903471477408475598460',
        height: 707395,
        microBlockHash: 'mh_2dUTfmwFc2ymeroB534giVwEvsa8d44Vf8SXtvy6GeHjdgQoHj',
        microBlockTime: 1671708830503n,
        updatedAt: new Date('2024-03-20 12:16:49.065'),
        pair: {
          id: 3,
          address: 'ct_22iY9F7hng23gN8awi4aGnLy54YSR41wztbqgQCquuLYvTiGcm',
          t0: 17,
          t1: 22,
          synchronized: true,
        },
      };

      mockPairLiquidityInfoHistoryService.getAllHistoryEntries.mockResolvedValue(
        [historyEntry1, historyEntry2],
      );

      // Call route
      const result = await request(app.getHttpServer()).get(
        '/history/liquidity',
      );

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryService.getAllHistoryEntries,
      ).toHaveBeenCalledWith(
        100,
        0,
        OrderQueryEnum.asc,
        undefined,
        undefined,
        undefined,
        undefined,
      );
      expect(result.status).toBe(200);
      expect(result.body).toEqual([
        {
          pairAddress: historyEntry1.pair.address,
          liquidityInfo: {
            totalSupply: historyEntry1.totalSupply,
            reserve0: historyEntry1.reserve0,
            reserve1: historyEntry1.reserve1,
          },
          height: historyEntry1.height,
          microBlockHash: historyEntry1.microBlockHash,
          microBlockTime: historyEntry1.microBlockTime.toString(),
        },
        {
          pairAddress: historyEntry2.pair.address,
          liquidityInfo: {
            totalSupply: historyEntry2.totalSupply,
            reserve0: historyEntry2.reserve0,
            reserve1: historyEntry2.reserve1,
          },
          height: historyEntry2.height,
          microBlockHash: historyEntry2.microBlockHash,
          microBlockTime: historyEntry2.microBlockTime.toString(),
        },
      ]);
    });

    it('should parse all query params correctly', async () => {
      // Mocks
      mockPairLiquidityInfoHistoryService.getAllHistoryEntries.mockResolvedValue(
        [],
      );

      // Call route
      const result = await request(app.getHttpServer()).get(
        '/history/liquidity?limit=50&offset=50&order=desc&pair-address=ct_22iY9&height=912485&from-block-time=1709027642807&to-block-time=1709027642807',
      );

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryService.getAllHistoryEntries,
      ).toHaveBeenCalledWith(
        50,
        50,
        OrderQueryEnum.desc,
        'ct_22iY9',
        912485,
        1709027642807n,
        1709027642807n,
      );
      expect(result.status).toBe(200);
      expect(result.body).toEqual([]);
    });

    it('should validate limit query param correctly', async () => {
      // Call route
      const result = await request(app.getHttpServer()).get(
        '/history/liquidity?limit=xyz',
      );

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryService.getAllHistoryEntries,
      ).toHaveBeenCalledTimes(0);
      expect(result.status).toBe(400);
    });

    it('should validate offset query param correctly', async () => {
      // Call route
      const result = await request(app.getHttpServer()).get(
        '/history/liquidity?offset=xyz',
      );

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryService.getAllHistoryEntries,
      ).toHaveBeenCalledTimes(0);
      expect(result.status).toBe(400);
    });

    it('should validate order query param correctly', async () => {
      // Call route
      const result = await request(app.getHttpServer()).get(
        '/history/liquidity?order=xyz',
      );

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryService.getAllHistoryEntries,
      ).toHaveBeenCalledTimes(0);
      expect(result.status).toBe(400);
    });

    it('should validate height query param correctly', async () => {
      // Call route
      const result = await request(app.getHttpServer()).get(
        '/history/liquidity?height=xyz',
      );

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryService.getAllHistoryEntries,
      ).toHaveBeenCalledTimes(0);
      expect(result.status).toBe(400);
    });

    it('should validate from-block-time query param correctly', async () => {
      // Call route
      const result = await request(app.getHttpServer()).get(
        '/history/liquidity?from-block-time=xyz',
      );

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryService.getAllHistoryEntries,
      ).toHaveBeenCalledTimes(0);
      expect(result.status).toBe(400);
    });

    it('should validate to-block-time query param correctly', async () => {
      // Call route
      const result = await request(app.getHttpServer()).get(
        '/history/liquidity?to-block-time=xyz',
      );

      // Assertions
      expect(
        mockPairLiquidityInfoHistoryService.getAllHistoryEntries,
      ).toHaveBeenCalledTimes(0);
      expect(result.status).toBe(400);
    });
  });
});
