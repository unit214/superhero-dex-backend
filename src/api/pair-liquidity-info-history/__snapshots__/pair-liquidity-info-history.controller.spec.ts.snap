// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairLiquidityInfoHistoryController GET /history should return history entries and use default values for empty params 1`] = `
[
  {
    "aeUsdPrice": "0.050559",
    "delta0UsdValue": "2.556212481e-18",
    "delta1UsdValue": "2.556212481e-18",
    "deltaReserve0": "1000",
    "deltaReserve1": "1000",
    "height": 100001,
    "logIndex": 1,
    "microBlockHash": "mh_entry1",
    "microBlockTime": "1000000000001",
    "pairAddress": "ct_pair1",
    "reserve0": "1000",
    "reserve0Usd": "2.556212481e-18",
    "reserve1": "1000",
    "reserve1Usd": "2.556212481e-18",
    "senderAccount": "",
    "token0AePrice": "0.050559",
    "token1AePrice": "0.050559",
    "transactionHash": "th_entry1",
    "transactionIndex": "100001",
    "txUsdFee": "7.668637443e-21",
    "type": "PairMint",
  },
  {
    "aeUsdPrice": "0.050559",
    "delta0UsdValue": "2.556212481e-18",
    "delta1UsdValue": "2.556212481e-18",
    "deltaReserve0": "1000",
    "deltaReserve1": "1000",
    "height": 300003,
    "logIndex": 1,
    "microBlockHash": "mh_entry3",
    "microBlockTime": "3000000000003",
    "pairAddress": "ct_pair2",
    "reserve0": "1000",
    "reserve0Usd": "2.556212481e-18",
    "reserve1": "1000",
    "reserve1Usd": "2.556212481e-18",
    "senderAccount": "",
    "token0AePrice": "0.050559",
    "token1AePrice": "0.050559",
    "transactionHash": "th_entry3",
    "transactionIndex": "300003",
    "txUsdFee": "7.668637443e-21",
    "type": "PairMint",
  },
]
`;
